File: composeApp/src/androidMain/kotlin/com/muralex/exploramus/ui/adaptive/FormFactor.kt
package com.muralex.exploramus.ui.adaptive

import android.annotation.SuppressLint
import android.content.res.Configuration
import androidx.compose.material3.adaptive.currentWindowAdaptiveInfo
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.platform.LocalConfiguration
import androidx.window.core.layout.WindowSizeClass

val LocalFormFactor = staticCompositionLocalOf<FormFactor> {
    error("LocalFormFactor not provided — wrap with CompositionLocalProvider at root")
}


data class FormFactor(
    val widthType: WidthType,
    val heightType: HeightType,
    val orientation: ScreenOrientation
)

enum class WidthType {
    COMPACT,
    MEDIUM,
    EXPANDED
}

enum class HeightType {
    COMPACT,
    MEDIUM,
    EXPANDED
}

enum class ScreenOrientation {
    PORTRAIT,
    LANDSCAPE
}

val FormFactor.isPhone get() = this.widthType == WidthType.COMPACT
val FormFactor.isLarge get() = this.widthType != WidthType.COMPACT && this.heightType != HeightType.COMPACT
val FormFactor.isCompact get() = this.widthType == WidthType.COMPACT || this.heightType == HeightType.COMPACT
val FormFactor.isLandscape get() = this.orientation == ScreenOrientation.LANDSCAPE
val FormFactor.isLargeOrLandscape get() = isLarge || isLandscape
val FormFactor.useBottomBar get() = this.widthType == WidthType.COMPACT
val FormFactor.useNavRail get() = this.widthType == WidthType.MEDIUM
val FormFactor.useDrawer get() = this.widthType == WidthType.EXPANDED
val FormFactor.applyTopBarScroll get() = this.heightType == HeightType.COMPACT

val FormFactor.isCompactHeight get() = this.heightType == HeightType.COMPACT

@SuppressLint("ConfigurationScreenWidthHeight")
@Composable
fun rememberFormFactor(): FormFactor {
    val adaptiveInfo = currentWindowAdaptiveInfo()
    val configuration = LocalConfiguration.current

    return remember(adaptiveInfo) {
        val isCompactWidth = !adaptiveInfo.windowSizeClass
            .isWidthAtLeastBreakpoint(WindowSizeClass.WIDTH_DP_MEDIUM_LOWER_BOUND)

        val isExpandedWidth = adaptiveInfo.windowSizeClass
            .isWidthAtLeastBreakpoint(WindowSizeClass.WIDTH_DP_EXPANDED_LOWER_BOUND)

        val isCompactHeight = !adaptiveInfo.windowSizeClass
            .isHeightAtLeastBreakpoint(WindowSizeClass.HEIGHT_DP_MEDIUM_LOWER_BOUND)

        val isExpandedHeight = adaptiveInfo.windowSizeClass
            .isHeightAtLeastBreakpoint(WindowSizeClass.HEIGHT_DP_EXPANDED_LOWER_BOUND)

        val widthType = when {
            isCompactWidth -> WidthType.COMPACT
            isExpandedWidth -> WidthType.EXPANDED
            else -> WidthType.MEDIUM
        }

        val heightType = when {
            isCompactHeight  -> HeightType.COMPACT
            isExpandedHeight -> HeightType.EXPANDED
            else -> HeightType.MEDIUM
        }

        val screenOrientation =
            if (configuration.orientation == Configuration.ORIENTATION_LANDSCAPE)
                ScreenOrientation.LANDSCAPE
            else
                ScreenOrientation.PORTRAIT

        FormFactor(
            widthType = widthType,
            heightType = heightType,
            orientation = screenOrientation
        ).also{
            println("Formfactor: $it , width ${ configuration.screenWidthDp}, height ${configuration.screenHeightDp} ")
        }
    }
}

File: iosApp/iosApp/design/adaptive/FormFactor.swift


import SwiftUI


struct FormFactor {
    let widthType: SizeType
    let heightType: SizeType
    let orientation: ScreenOrientation
    let deviceType: DeviceType
    
    var isPhone: Bool { deviceType == .phone }
    var isTablet: Bool { deviceType == .tablet }
}

extension FormFactor {
    static func defaultValue() -> FormFactor {
        return FormFactor(
            widthType: .compact,
            heightType: .medium,
            orientation: .portrait,
            deviceType: .phone //deviceType
        )
    }
    
    static func from(
        horizontal: UserInterfaceSizeClass?,
        vertical: UserInterfaceSizeClass?,
        size: CGSize
    ) -> FormFactor {
        
        let deviceType: DeviceType = UIDevice.current.userInterfaceIdiom == .pad ? .tablet : .phone
        
        let widthType: SizeType = {
            if horizontal == .compact { return .compact }
            if horizontal == .regular && vertical == .regular { return .expanded }
            return .medium
        }()
        
        let heightType: SizeType = {
            if vertical == .compact { return .compact }
            if vertical == .regular && horizontal == .regular { return .expanded }
            return .medium
        }()

        let orientation: ScreenOrientation = size.width > size.height ? .landscape : .portrait

        return FormFactor(
            widthType: widthType,
            heightType: heightType,
            orientation: orientation,
            deviceType: deviceType
        )
    }
}


enum SizeType {
    case compact, medium, expanded
}

enum ScreenOrientation {
    case portrait, landscape
}

enum DeviceType {
    case phone, tablet
}


